segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/SegmentMetadata.java

```diff
@@ -0,0 +1,334 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.segmentstore.storage.metadata;
+
+import com.google.common.base.Preconditions;
+import io.pravega.common.ObjectBuilder;
+import io.pravega.common.io.serialization.RevisionDataInput;
+import io.pravega.common.io.serialization.RevisionDataOutput;
+import io.pravega.common.io.serialization.VersionedSerializer;
+import lombok.Builder;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+import java.io.IOException;
+
+/**
+ * Represents segment metadata.
+ Following metadata is stored.
+ * <ul>
+ *     <li>Name of the segment.</li>
+ *     <li>Length of the segment.</li>
+ *     <li>Epoch of the container that last owned it.</li>
+ *     <li>Start offset of the segment. This is offset of the first byte available for read.</li>
+ *     <li>Status bit flags.</li>
+ *     <li>Maximum Rolling length of the segment.</li>
+ *     <li>Name of the first chunk.</li>
+ *     <li>Name of the last chunk.</li>
+ *     <li>Offset corresponding to the the first byte of the first chunk.This is NOT the same as start offset of the segment.
+ *      With arbitrary truncates the effective start offset might be in the middle of the first chunk. Byte at this offset may not be available for read.</li>
+ *     <li>Offset of the first byte of the last chunk.</li>
+ * </ul>
+ */
+@Data
+@Builder(toBuilder = true)
+@EqualsAndHashCode(callSuper = true)
+public class SegmentMetadata extends StorageMetadata {
+    /**
+     * Flag to indicate whether the segment is active or not.
+     */
+    private static final int ACTIVE = 0x0001;
+
+    /**
+     * Flag to indicate whether the segment is sealed or not.
+     */
+    private static final int SEALED = 0x0002;
+
+    /**
+     * Flag to indicate whether the segment is deleted or not.
+     */
+    private static final int DELETED = 0x0004;
+
+    /**
+     * Flag to indicate whether followup actions (like adding new chunks) after ownership changes are needed or not.
+     */
+    private static final int OWNERSHIP_CHANGED = 0x0008;
+
+    /**
+     * Flag to indicate whether the segment is storage system segment.
+     */
+    private static final int SYSTEM_SEGMENT = 0x0010;
+
+    /**
+     * Name of the segment.
+     */
+    private final String name;
+
+    /**
+     * Length of the segment.
+     */
+    private long length;
+
+    /**
+     * Number of chunks.
+     */
+    private int chunkCount;
+
+    /**
+     * Start offset of the segment. This is offset of the first byte available for read.
+     */
+    private long startOffset;
+
+    /**
+     * Status bit flags.
+     */
+    private int status;
```

Comment:
Just for curiosity, why are we packing the status in a bitset ? is it for saving resources ?
is it some form of convention in Pravega project ?

Reply:
Yes - it is to conserve amount of data store and also memory footprint. Each boolean takes space. But putting it in status helps save at least there.

Unlike C#, unfortunately Java enums are not very flexible or that useful here. (https://docs.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=netcore-3.1)

Reply:
In fact if I don't have more than 8 flags, I'll change this field to byte.